[Debug] ASSERTION FAILED: !isSuspended() on `fast/css/view-transitions-zoom.html`

obyknovenius opened the pull request:
#### 9e7f7fec1b29c528a4340da9e76c9583c7c59bc0
<pre>
[Debug] ASSERTION FAILED: !isSuspended() on `fast/css/view-transitions-zoom.html`
<a href="https://bugs.webkit.org/show_bug.cgi?id=279589">https://bugs.webkit.org/show_bug.cgi?id=279589</a>

Reviewed by Tim Nguyen.

Make sure that the script execution context is not stopped before
evaluating the status of the promise when it&apos;s settled.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):

Canonical link: <a href="https://commits.webkit.org/283996@main">https://commits.webkit.org/283996@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb4731bb64170f8806263f7de85e8471020ab86c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68000 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72059 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18961 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54349 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12768 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71067 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58769 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34818 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16186 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17502 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62044 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16527 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73755 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11971 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15805 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61810 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12009 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61824 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15117 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9724 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3340 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43192 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44266 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45461 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44008 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -187,6 +187,9 @@ void ViewTransition::skipViewTransition(ExceptionOr<JSC::JSValue>&& reason)
     }
 
     m_updateCallbackDone.first->whenSettled([this, protectedThis = Ref { *this }] {
+        if (m_updateCallbackDone.first->isSuspended())
```
Checking if the ViewTransition object `isStopped()` is the more common pattern.

kmiller68 replied:
Does it work to push this check into `whenPromiseIsSettled`? I assume it would be annoying/error-prone to remember that documents can get detached whenever you handle a promise in WebCore.

obyknovenius replied:
> Checking if the ViewTransition object `isStopped()` is the more common pattern.

ViewTransition doesn't have `isStopped()` method. Do you mean `ActiveDOMObject::isContextStopped()`?

obyknovenius replied:
> Does it work to push this check into `whenPromiseIsSettled`? I assume it would be annoying/error-prone to remember that documents can get detached whenever you handle a promise in WebCore.

What are you suggesting? Don't call the callback at all if the promise is suspended? Could there be cases where we do something in the callback regardless of the status of the promise?

obyknovenius committed:
[Debug] ASSERTION FAILED: !isSuspended() on `fast/css/view-transitions-zoom.html`
https://bugs.webkit.org/show_bug.cgi?id=279589

Reviewed by Tim Nguyen.

Make sure that the script execution context is not stopped before
evaluating the status of the promise when it's settled.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::skipViewTransition):

Canonical link: https://commits.webkit.org/283996@main
